### PR TITLE
SIP002 and SIP003 support.

### PR DESCRIPTION
**Pull request by rwasef1830**
Hello,
This is a first draft WIP to get feedback on my approach. This pull request implements SIP002 url (parsing and generation for QR code with backward-compatibility of accepting old-style urls and generating old-style urls when there is no plugin for the config), and SIP003 plugin interface for shadowsocks-windows.

- SIP002 has unit tests and proven to work.
- SIP003 needs a little bit of testing, would appreciate feedback on the approach and fresh eyes to see if I made any mistakes.

Hope to eventually get this merged soon :-)

### CODE REVIEW

**Comment by rwasef1830** [comment]
Fixed it enough, seems to work now. Tested with simple-obfs but is it normal to have plugin process for each connection ? Waiting for feedback, thanks in advance.

**Comment by rwasef1830** [comment]
I will implement a better approach for this in the coming days (single process for plugin, and cleaner implementation).

**Comment by celeron533** [comment]
Thank you @rwasef1830 
Since this is a significant change and functional modification, need more developers to do the review.

**Comment by rwasef1830** [comment]
New cleaner approach done. Now it is a single plugin process per server and it is automatically restarted if it crashes or is killed. Process launch is protected by lock.

Waiting for review and merge ;-)

**Comment by rwasef1830** [comment]
Turns out that it is not possible to compile simple-obfs using mingw-w64 (because of posix semantics, error in sys/select.h), so I redid build on latest master and bundled the needed msys dlls in the archive.

64-bit simple-obfs https://github.com/shadowsocks/simple-obfs/commit/1126971c0cc4c2ee93046689440650cb56376417 msys2 build. Should run on any 64-bit Windows 7 and higher.
[simple-obfs-0.0.3-1126971.zip](https://github.com/shadowsocks/shadowsocks-windows/files/1217741/simple-obfs-0.0.3-1126971.zip)

Still waiting for feedback and/or merge ;-)

**Comment by chenshaoju** [comment]
The simple-obfs-0.0.3-1126971.zip look's work for me.

Saved,Thanks. :+1:

**Comment by rwasef1830** [comment]
Rebased on lastest master. I hope this get merged soon, can't wait to use this functionality :-)

**Comment by celeron533** [comment]
I saw you added `ProcessStartInfo.Environment`. Will all the plugins follow this standard in the future?
```
    Environment =
    {
        ["SS_REMOTE_HOST"] = serverAddress,
        ["SS_REMOTE_PORT"] = serverPort.ToString(),
        ["SS_PLUGIN_OPTIONS"] = pluginOpts
    }
```

**Comment by rwasef1830** [comment]
Yes, this is described in the SIP003 standard https://github.com/shadowsocks/shadowsocks-org/issues/28 and https://shadowsocks.org/en/spec/Plugin.html

Any SIP003 plugin to work with shadowsocks must read from these environment vars, so when I start the plugin, I pass them in ProcessStartInfo.

**Comment by rwasef1830** [comment]
Made a force push to fix a tiny embarrassing one line null check in StopPlugins() (forgot to check if the server plugin was null :-D).

File: shadowsocks-csharp/Proxy/Sip003Plugin.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        }
+
+        private Sip003Plugin(string plugin, string pluginOpts, string serverAddress, int serverPort)
+        {
+            if (plugin == null) throw new ArgumentNullException(nameof(plugin));
+            if (string.IsNullOrWhiteSpace(serverAddress))
+            {
+                throw new ArgumentException("Value cannot be null or whitespace.", nameof(serverAddress));
+            }
+            if ((uint)serverPort != serverPort)
```
Comment (celeron533) [inline shadowsocks-csharp/Proxy/Sip003Plugin.cs]: Is it necessary to check the negative port number?
Port number should between 1 to 65535.
Server port in config file should have been verified when user added the new entry.

Comment (rwasef1830) [inline shadowsocks-csharp/Proxy/Sip003Plugin.cs]: Ah a silly mistake, I was intending to cast to ushort to verify range ... etc.
I don't like depending on checking only in higher levels of code in general, I consider each method should check its inputs (especially if there is negligible performance cost).

**Comment by rwasef1830** [comment]
@celeron533 pushed a fix to the range check and moving the Sip003Plugin class.

**Comment by madeye** [comment]
I thinks this pull request is ready to merge?

**Comment by rwasef1830** [comment]
@madeye @celeron533 For completeness sake, there are a few strings added to the GUI and I didn't add translations for them.

**Comment by celeron533** [comment]
Added a notification message on Forward Proxy form.

preview: https://github.com/celeron533/shadowsocks-windows/commit/05df4498260df19e41265b02614f992679111bc0

image

image

**Comment by nerdDan** [comment]
do i have to compile `simple-obfs` on my own to use it?

**Comment by celeron533** [comment]
@nerdDan  #1379

**Pull request merged**